internal/authz/authz.go

```diff
@@ -0,0 +1,237 @@
+	WriteAssertions         = "WriteAssertions"
+	ReadAssertions          = "ReadAssertions"
+	WriteAuthorizationModel = "WriteAuthorizationModel"
+	ListStores              = "ListStores"
+	CreateStore             = "CreateStore"
+	GetStore                = "GetStore"
+	DeleteStore             = "DeleteStore"
+	Expand                  = "Expand"
+	ReadChanges             = "ReadChanges"
+
+	// Relations.
+	CanCallReadAuthorizationModels  = "can_call_read_authorization_models"
+	CanCallRead                     = "can_call_read"
+	CanCallWrite                    = "can_call_write"
+	CanCallListObjects              = "can_call_list_objects"
+	CanCallCheck                    = "can_call_check"
+	CanCallListUsers                = "can_call_list_users"
+	CanCallWriteAssertions          = "can_call_write_assertions"
+	CanCallReadAssertions           = "can_call_read_assertions"
+	CanCallWriteAuthorizationModels = "can_call_write_authorization_models"
+	CanCallListStores               = "can_call_list_stores"
+	CanCallCreateStore              = "can_call_create_stores"
+	CanCallGetStore                 = "can_call_get_store"
+	CanCallDeleteStore              = "can_call_delete_store"
+	CanCallExpand                   = "can_call_expand"
+	CanCallReadChanges              = "can_call_read_changes"
+
+	StoreType       = "store"
+	ApplicationType = "application"
+	SystemType      = "system"
+	RootSystemID    = "fga"
+)
+
+var (
+	ErrorResponse = &openfgav1.ForbiddenResponse{Code: 403, Message: "the principal is not authorized to perform the action"}
+)
+
+type Config struct {
+	StoreID string
+	ModelID string
+}
+
+type Authorizer struct {
+	config *Config
+	server ServerInterface
+	logger logger.Logger
+}
+
+// NewAuthorizer creates a new authorizer.
+func NewAuthorizer(config *Config, server ServerInterface, logger logger.Logger) *Authorizer {
+	return &Authorizer{
+		config: config,
+		server: server,
+		logger: logger,
+	}
+}
+
+func (a *Authorizer) getRelation(apiMethod string) (string, error) {
+	switch apiMethod {
+	case ReadAuthorizationModel, ReadAuthorizationModels:
+		return CanCallReadAuthorizationModels, nil
+	case Read:
+		return CanCallRead, nil
+	case Write:
+		return CanCallWrite, nil
+	case ListObjects, StreamedListObjects:
+		return CanCallListObjects, nil
+	case Check:
+		return CanCallCheck, nil
+	case ListUsers:
+		return CanCallListUsers, nil
+	case WriteAssertions:
+		return CanCallWriteAssertions, nil
+	case ReadAssertions:
+		return CanCallReadAssertions, nil
+	case WriteAuthorizationModel:
+		return CanCallWriteAuthorizationModels, nil
+	case ListStores:
+		return CanCallListStores, nil
+	case CreateStore:
+		return CanCallCreateStore, nil
+	case GetStore:
+		return CanCallGetStore, nil
+	case DeleteStore:
+		return CanCallDeleteStore, nil
+	case Expand:
+		return CanCallExpand, nil
+	case ReadChanges:
+		return CanCallReadChanges, nil
+	default:
+		return "", fmt.Errorf("unknown api method: %s", apiMethod)
```

Comment:
This error needs to be either a proper error with status code or it needs to be handled (i.e. transformed) to a proper error by the calling code, somewhere

Reply:
Can we instead create a custom Authorization error? Something like:
```go
var UnknownApiMethod = erros.New("unknown API method")

type AuthorizationError struct {
    Err error
}

func (e *AuthorizationError) Error() string {
    fmt.Sprintf("error authorizing request: %s", e.Err)
}

func (e *AuthorizationError) Unwrap() error {
    return e.Err
}
```
Put together:
```go
err := AuthorizationError{
    Err: UnknownApiMethod
}
```
We can then use `errors.Is` to determine what type of error we are dealing with programatically.